File: python/src/billing/application/dto.py
"""ユースケースの入出力。

presentation 層の Pydantic モデルをここに持ち込まない。ユースケースは HTTP 経由でも
CLI からでもバッチからでも呼ばれうるので、その形は「FastAPI が読みやすい形」であっては
ならない。素の dataclass にしておけば、どこから呼んでも同じ。
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime

from billing.domain.invoice import Invoice
from billing.domain.plan import Plan
from billing.domain.proration import Proration
from billing.domain.subscription import Subscription


@dataclass(frozen=True, slots=True)
class SubscribeCommand:
    customer_id: str
    plan_id: str
    trial_days: int | None = None


@dataclass(frozen=True, slots=True)
class ChangePlanCommand:
    subscription_id: str
    new_plan_id: str
    idempotency_key: str


@dataclass(frozen=True, slots=True)
class CancelCommand:
    subscription_id: str
    immediately: bool = False


@dataclass(frozen=True, slots=True)
class PaymentNotification:
    """決済代行から届く支払い結果の通知。"""

    invoice_id: str
    succeeded: bool
    provider_reference: str | None = None
    failure_reason: str | None = None


@dataclass(frozen=True, slots=True)
class MoneyView:
    amount: int
    currency: str


@dataclass(frozen=True, slots=True)
class PlanView:
    id: str
    name: str
    price: MoneyView
    interval: str

    @classmethod
    def of(cls, plan: Plan) -> PlanView:
        return cls(
            id=plan.id,
            name=plan.name,
            price=MoneyView(plan.price.amount, plan.price.currency),
            interval=str(plan.interval),
        )


@dataclass(frozen=True, slots=True)
class SubscriptionView:
    id: str
    customer_id: str
    plan_id: str
    status: str
    current_period_start: datetime
    current_period_end: datetime
    cancel_at_period_end: bool
    trial_end: datetime | None

    @classmethod
    def of(cls, subscription: Subscription) -> SubscriptionView:
        return cls(
            id=subscription.id,
            customer_id=subscription.customer_id,
            plan_id=subscription.plan_id,
            status=str(subscription.status),
            current_period_start=subscription.current_period.start,
            current_period_end=subscription.current_period.end,
            cancel_at_period_end=subscription.cancel_at_period_end,
            trial_end=subscription.trial_end,
        )


@dataclass(frozen=True, slots=True)
class InvoiceView:
    id: str
    subscription_id: str
    status: str
    total: MoneyView
    lines: list[tuple[str, MoneyView]]
    issued_at: datetime | None
    paid_at: datetime | None

    @classmethod
    def of(cls, invoice: Invoice) -> InvoiceView:
        return cls(
            id=invoice.id,
            subscription_id=invoice.subscription_id,
            status=str(invoice.status),
            total=MoneyView(invoice.total.amount, invoice.total.currency),
            lines=[
                (line.description, MoneyView(line.amount.amount, line.amount.currency))
                for line in invoice.lines
            ],
            issued_at=invoice.issued_at,
            paid_at=invoice.paid_at,
        )


@dataclass(frozen=True, slots=True)
class ProrationView:
    credit: MoneyView
    charge: MoneyView
    net: MoneyView

    @classmethod
    def of(cls, proration: Proration) -> ProrationView:
        return cls(
            credit=MoneyView(proration.credit.amount, proration.credit.currency),
            charge=MoneyView(proration.charge.amount, proration.charge.currency),
            net=MoneyView(proration.net.amount, proration.net.currency),
        )


@dataclass(frozen=True, slots=True)
class SubscribeResult:
    subscription: SubscriptionView
    invoice: InvoiceView | None
    payment_failed: bool = False


@dataclass(frozen=True, slots=True)
class ChangePlanResult:
    subscription: SubscriptionView
    proration: ProrationView
    invoice: InvoiceView
    """プラン変更は差額が 0 以下でも必ず請求書を残すので、常に存在する。"""


@dataclass(frozen=True, slots=True)
class RenewalReport:
    """バッチ 1 回ぶんの結果。"""

    renewed: int = 0
    invoiced: int = 0
    payment_failed: int = 0
    terminated: int = 0
    canceled_for_nonpayment: int = 0

File: go/internal/adapter/http/dto.go
package http

import (
	"time"

	"github.com/Yutarotaro/clean-architecture-billing/go/internal/usecase"
)

// JSON の構造体はこのファイルの外に出さない。ユースケースは usecase パッケージの
// View だけを返し、それをここで 1 回変換する。手間を引き受ける代わりに、
// 「API の互換性のためにフィールド名を変えたい」がドメインに波及しなくなる。

type moneyJSON struct {
	Amount   int64  `json:"amount"`
	Currency string `json:"currency"`
}

func toMoneyJSON(v usecase.MoneyView) moneyJSON {
	return moneyJSON{Amount: v.Amount, Currency: v.Currency}
}

type planJSON struct {
	ID       string    `json:"id"`
	Name     string    `json:"name"`
	Price    moneyJSON `json:"price"`
	Interval string    `json:"interval"`
}

func toPlanJSON(v usecase.PlanView) planJSON {
	return planJSON{ID: v.ID, Name: v.Name, Price: toMoneyJSON(v.Price), Interval: v.Interval}
}

type subscriptionJSON struct {
	ID                 string     `json:"id"`
	CustomerID         string     `json:"customer_id"`
	PlanID             string     `json:"plan_id"`
	Status             string     `json:"status"`
	CurrentPeriodStart time.Time  `json:"current_period_start"`
	CurrentPeriodEnd   time.Time  `json:"current_period_end"`
	CancelAtPeriodEnd  bool       `json:"cancel_at_period_end"`
	TrialEnd           *time.Time `json:"trial_end"`
}

func toSubscriptionJSON(v usecase.SubscriptionView) subscriptionJSON {
	return subscriptionJSON{
		ID:                 v.ID,
		CustomerID:         v.CustomerID,
		PlanID:             v.PlanID,
		Status:             v.Status,
		CurrentPeriodStart: v.CurrentPeriodStart,
		CurrentPeriodEnd:   v.CurrentPeriodEnd,
		CancelAtPeriodEnd:  v.CancelAtPeriodEnd,
		TrialEnd:           v.TrialEnd,
	}
}

type invoiceLineJSON struct {
	Description string    `json:"description"`
	Amount      moneyJSON `json:"amount"`
}

type invoiceJSON struct {
	ID             string            `json:"id"`
	SubscriptionID string            `json:"subscription_id"`
	Status         string            `json:"status"`
	Total          moneyJSON         `json:"total"`
	Lines          []invoiceLineJSON `json:"lines"`
	IssuedAt       *time.Time        `json:"issued_at"`
	PaidAt         *time.Time        `json:"paid_at"`
}

func toInvoiceJSON(v usecase.InvoiceView) invoiceJSON {
	lines := make([]invoiceLineJSON, 0, len(v.Lines))
	for _, line := range v.Lines {
		lines = append(lines, invoiceLineJSON{
			Description: line.Description,
			Amount:      toMoneyJSON(line.Amount),
		})
	}
	return invoiceJSON{
		ID:             v.ID,
		SubscriptionID: v.SubscriptionID,
		Status:         v.Status,
		Total:          toMoneyJSON(v.Total),
		Lines:          lines,
		IssuedAt:       v.IssuedAt,
		PaidAt:         v.PaidAt,
	}
}

func toInvoiceJSONPtr(v *usecase.InvoiceView) *invoiceJSON {
	if v == nil {
		return nil
	}
	converted := toInvoiceJSON(*v)
	return &converted
}

type prorationJSON struct {
	Credit moneyJSON `json:"credit"`
	Charge moneyJSON `json:"charge"`
	Net    moneyJSON `json:"net"`
}

func toProrationJSON(v usecase.ProrationView) prorationJSON {
	return prorationJSON{
		Credit: toMoneyJSON(v.Credit),
		Charge: toMoneyJSON(v.Charge),
		Net:    toMoneyJSON(v.Net),
	}
}

type subscribeRequest struct {
	CustomerID string `json:"customer_id"`
	PlanID     string `json:"plan_id"`
	TrialDays  int    `json:"trial_days"`
}

type subscribeResponse struct {
	Subscription  subscriptionJSON `json:"subscription"`
	Invoice       *invoiceJSON     `json:"invoice"`
	PaymentFailed bool             `json:"payment_failed"`
}

type changePlanRequest struct {
	NewPlanID string `json:"new_plan_id"`
}

type changePlanResponse struct {
	Subscription subscriptionJSON `json:"subscription"`
	Proration    prorationJSON    `json:"proration"`
	// プラン変更は差額が 0 以下でも必ず請求書を残すので、常に存在する。
	Invoice invoiceJSON `json:"invoice"`
}

type cancelRequest struct {
	Immediately bool `json:"immediately"`
}

type paymentWebhookRequest struct {
	InvoiceID         string `json:"invoice_id"`
	Succeeded         bool   `json:"succeeded"`
	ProviderReference string `json:"provider_reference"`
	FailureReason     string `json:"failure_reason"`
}

type renewalReportJSON struct {
	Renewed               int `json:"renewed"`
	Invoiced              int `json:"invoiced"`
	PaymentFailed         int `json:"payment_failed"`
	Terminated            int `json:"terminated"`
	CanceledForNonpayment int `json:"canceled_for_nonpayment"`
}

type errorJSON struct {
	Error  string `json:"error"`
	Detail string `json:"detail"`
}
